### ISSUE TITLE
Layer 2 Annoucment outbound traffic silently dropped with firewalld

### ISSUE THREAD

**Issue by j-gooding** [cilium/cilium#26366]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

This may or may not be a bug, either way it could use some documentation. I'm trying to get cilium to work with metallb and firewalld on a bare metal on-perm setup. The long story short is that outbound traffic via l2 announced gateway api service gets silently dropped and never makes it to the external client.

Overall software
1. rocky linux 9.2
2. firewalld is enabled
3. kubernetes v1.13.4
  - using kubeadm, containerd
5. metallb 0.13.10
6. gateway api 0.7.1 https://gateway-api.sigs.k8s.io/guides/#install-experimental-channel
7. kube-proxy replacement is strict
8. tunnel is vxlan

Networking is as follows:
1. Public network a.b.c.d/16
2. Private network for control plane 10.104.120.0/24
3. Pod Cidr 10.1.0.0/16
4. Service CIdr 10.0.0.0/16
5. There is only one node for proof of concept right now, untainted
9. The node has two NICs, eno0 and eno1
10. eno0 connects to a.b.c.d which is public and connected to the internet
11. eno0 is assigned a static address from a.b.c.d/16
12. eno1 is assigned a static address from 10.104.120.0/24
13. k8s api is bound to eno1 and 10.104.120.0/24, it is not bound to all interfaces as is the default

Firewalld is as follows:
```
internal
  target: default
  interfaces:
  sources: a.b.c.d/16
  services: ssh http https etc.
  protocols: icmp ipv6-icmp
  forward: yes
  masquerade: yes (cluster doesn't work without this)
trusted
  target: ACCEPT
  interfaces: eno1 (assigned static IP from 10.104.120.0/24)
  sources: 10.104.120.0/24 10.0.0.0/16 10.1.0.0.0/16 (cluster doesn't work unless I add pod and service cidrs)
  forward: yes
  masquerade: no (can be yes or no, appears to have no effect)
public
  target: default
  interfaces: eno0 (eno0 is assigned static address from a.b.c.d/16)
  forward: yes
  masquerade: no (setting this to yes if trusted is no, breaks the cluster, setting this to yes if trusted is also yes, the cluster works and appears to have no effect.)
```

1. I am L2 announcing a single service of cluster IP type with the cilium gateway API via MetalLB. 
2. metallb is announcing from eno0 only. 
3. metallb announcing from both eno0 and eno1 makes no difference.
4. Metallb/cilium creates the load balancer service and I have verified it announces properly.
5. I can curl the service cluster IP and external IP on the local node. 
6. I can curl the external IP from any other machine on the L2 network if the firewall is disabled.
7. If I enable firewalld I can curl the service cluster IP and external IP on the local node.
8. If I enable firewalld and attempt to curl the external IP from another client on the network, curl is able to establish a connection, however it hangs indefinitely because a response is never received.
9. Using the pod logs I can tell that the pod received the external client traffic and successfully sent a response.
10. Using hubble observe I can tell there is a slight difference in the flow between a successful response on the local node, and one that doesn't make it back to the client.

The response is getting lost or dropped somewhere for external clients. My guess is this is a NAT issue but I can't be certain. The packets do not appear in firewalld drop logs. As far as I can tell the response is just silently getting lost somewhere, or if it is being tracked it isn't obvious where to me.

### Cilium Version

cilium-cli: v0.14.7 compiled with go1.20.4 on linux/amd64
cilium image (default): v1.13.3
cilium image (stable): v1.13.4
cilium image (running): v1.13.4

### Kernel Version

5.14.0-284.11.1.el9_2.x86_64

### Kubernetes Version

clientVersion:
  buildDate: "2023-05-17T14:20:07Z"
  compiler: gc
  gitCommit: 7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647
  gitTreeState: clean
  gitVersion: v1.27.2
  goVersion: go1.20.4
  major: "1"
  minor: "27"
  platform: linux/amd64
kustomizeVersion: v5.0.1
serverVersion:
  buildDate: "2023-04-11T17:04:24Z"
  compiler: gc
  gitCommit: 1b4df30b3cdfeaba6024e81e559a6cd09a089d65
  gitTreeState: clean
  gitVersion: v1.27.0
  goVersion: go1.20.3
  major: "1"
  minor: "27"
  platform: linux/amd64

### Sysdump

Will send through private channel if it is actually needed. I'm not going to sanitize every file in the zip, so won't publicity upload it.

To recreate a minimally working example of the cluster:

1. Inspect the files and replace the dummy IP's or change them as needed
2. run kube admin init with kubeadm.yaml, skipping proxy for cilium strict mode.
3. Install the linked gateway API crd above.
4. Install metallb with helm into metallb-system namespace
5. kubectl apply -f metallb-config.yaml
6. Install cilium with cilium-config.yaml via helm
7. kubectl apply -f basic-http-gateway.yaml
8. kubectl apply -f echo-server.yaml

[setup.zip](https://github.com/cilium/cilium/files/11792723/setup.zip)

### Relevant log output

```shell
a.b.C.D is the k8s node a.b.c.d/16 eno0 address
a.b.C.E is the external clients a.b.c.d/16 address
a.b.C.F is the metallb announced address on the k8s node for the a.b.c.d/16 network

curl for successful local node:
curl a.b.C.F/?echo_body=from_local_node
"from_local_node"

hubble observe pod for successful local connection:

Jun 19 19:23:58.407: 10.0.0.104:35695 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: SYN)
Jun 19 19:23:58.407: 10.0.0.104:35695 (host) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-stack FORWARDED (TCP Flags: SYN, ACK)
Jun 19 19:23:58.407: 10.0.0.104:35695 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: ACK)
Jun 19 19:23:58.407: 10.0.0.104:35695 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Jun 19 19:23:58.407: 127.0.0.1:59630 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) http-request FORWARDED (HTTP/1.1 GET http://a.b.C.F/?echo_body=from_local_node)
Jun 19 19:23:58.409: 10.0.0.104:35695 (host) <> echoserver/echoserver-7cbd5dd75-j2z9t (ID:34251) pre-xlate-rev TRACED (TCP)
Jun 19 19:23:58.409: 10.0.0.104:35695 (host) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-stack FORWARDED (TCP Flags: ACK, PSH)
Jun 19 19:23:58.409: 127.0.0.1:59630 (ingress) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) http-response FORWARDED (HTTP/1.1 200 2ms (GET http://a.b.C.F/?echo_body=from_local_node))
Jun 19 19:24:03.413: 10.0.0.104:35695 (host) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-stack FORWARDED (TCP Flags: ACK, FIN)
Jun 19 19:24:03.414: 10.0.0.104:35695 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: ACK, FIN)

curl for external client:
curl -v a.b.C.F/?echo_body=extern_client_same_layer_2
*   Trying a.b.C.F:80...
* Connected to a.b.C.F (a.b.C.F) port 80 (#0)
> GET /?echo_body=extern_client_same_layer_2 HTTP/1.1
> Host: a.b.C.F
> User-Agent: curl/7.76.1
> Accept: */*
> 
^C

hubble observe for external client

Jun 19 19:26:49.203: 10.0.0.104:34943 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: SYN)
Jun 19 19:26:49.203: 10.0.0.104:34943 (host) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-stack FORWARDED (TCP Flags: SYN, ACK)
Jun 19 19:26:49.203: 10.0.0.104:34943 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: ACK)
Jun 19 19:26:49.204: 10.0.0.104:34943 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Jun 19 19:26:49.204: a.b.C.E:51322 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) http-request FORWARDED (HTTP/1.1 GET http://a.b.C.F/?echo_body=extern_client_same_layer_2)
Jun 19 19:26:49.205: 10.0.0.104:34943 (host) <> echoserver/echoserver-7cbd5dd75-j2z9t (ID:34251) pre-xlate-rev TRACED (TCP)
Jun 19 19:26:49.206: 10.0.0.104:34943 (host) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-stack FORWARDED (TCP Flags: ACK, PSH)
Jun 19 19:26:49.206: a.b.C.E:51322 (ingress) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) http-response FORWARDED (HTTP/1.1 200 2ms (GET http://a.b.C.F/?echo_body=extern_client_same_layer_2))
Jun 19 19:26:54.210: 10.0.0.104:34943 (host) <- echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-stack FORWARDED (TCP Flags: ACK, FIN)
Jun 19 19:26:54.210: 10.0.0.104:34943 (ingress) -> echoserver/echoserver-7cbd5dd75-j2z9t:80 (ID:34251) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
```

### Anything else?

I'm willing to do a lot to help debug this and get it working. I just don't know where to look or how to tell where the response is getting dropped so that I can implement a fix. 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by j-gooding** [comment]
I should add as a caveat if you do try to reproduce this on a bare metal machine or in a VM, you **MUST** reboot the machine after you make a change to firewalld, _especially_ if you change anything related to `masquerade` or `--direct` rules. No amount of reloading services will save you. `firewalld` and `cilium` will **_only_** work properly for testing and evaluation if you do a hard reboot.

**Comment by borkmann** [comment]
Cc @dylandreimerink

**Comment by dylandreimerink** [comment]
@borkmann I seems like this is more about the interaction between firewallD and Gateway API service than L2 announcements since MetalLB is used for the L2 part and our new native implementation has been disabled.

From the setup.zip:
```
[...]
l2announcements:
  enabled: false
[...]
```

I can't spot anything obviously wrong from the provided logs, but I also have little knowledge about how the iptable rules generated by firewallD can/do interact with L7 traffic

**Comment by j-gooding** [comment]
@dylandreimerink sorry, I realize now my title is a little ambiguous. The L2 announcement was referring to metallb l2 announcement. I wasn't able to use cilium l2 due to the current lb ipam limitations, hence using metallb and ciliums being disabled.

And yeah it could totally be gateway api (which is cilium in this case), but I don't know since I can't find where it's getting dropped and nothing obvious is logging it inside or outside the cluster.

**Comment by margamanterola** [comment]
Have you tried inspecting the traffic using cilium monitor and/or hubble? Do you know where the traffic is getting dropped?

**Comment by j-gooding** [comment]
Hi @margamanterola I have tried. The results of the Hubble monitor are included in the original post above. I saw no meaningful difference between the two with firewalld enabled and disabled. Same with cilium monitor. tcpdump was identical on both nodes up to the point of the response never making it out of the originating node. I have no idea where the traffic is getting dropped. At this point if I had to take a guess it is somehow related to cilium_host and cilium_net. I think there is a missing rule between cilium rules added to nftables and firewalld nftable rules, but yeah, I can't say for sure, and I don't know how to debug it further. After further testing I think it is an issue with cilium itself, not the Gateway API, not metallb l2, not cilium ingress or nginx ingress, not cilium l2. I've tried all those things and it's the same results.
